File: 235-lowest-common-ancestor-of-a-binary-search-tree/lowest-common-ancestor-of-a-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        # Base case: if root is None, return None
        if not root:
            return None
        
        # If the current node is either p or q, return the current node
        if root == p or root == q:
            return root
        
        # Recur for the left and right subtree
        left = self.lowestCommonAncestor(root.left, p, q)
        right = self.lowestCommonAncestor(root.right, p, q)
        
        # If both left and right are non-null, it means p and q are found in
        # different subtrees, so the current node is their LCA
        if left and right:
            return root
        
        # Otherwise, return the non-null value between left and right
        return left if left else right

File: 235-lowest-common-ancestor-of-a-binary-search-tree/lowest-common-ancestor-of-a-binary-search-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */

/**
 * @param {TreeNode} root
 * @param {TreeNode} p
 * @param {TreeNode} q
 * @return {TreeNode}
 */
var lowestCommonAncestor = function(root, p, q) {
    var dfs=(node)=>{
        if(!node)return null;
        if(node ==p || node == q)return node
        let left = dfs(node.left)
        let right = dfs(node.right)
        if(left && right )return node
        return left ? left : right
    }

    return dfs(root)
};